android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/WifiUrlDeviceDiscoverer.java

```diff
@@ -0,0 +1,129 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.physical_web.physicalweb;
+
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.net.wifi.p2p.WifiP2pDevice;
+import android.net.wifi.p2p.WifiP2pDeviceList;
+import android.net.wifi.p2p.WifiP2pManager;
+import android.net.wifi.p2p.WifiP2pManager.Channel;
+import android.net.wifi.p2p.WifiP2pManager.PeerListListener;
+import android.util.Log;
+
+/**
+ * This class is for using WifiDirect to discover Physical Web
+ * Devices through peer discovery.
+ */
+class WifiUrlDeviceDiscoverer extends UrlDeviceDiscoverer {
+  private static final String TAG = "WifiUrlDeviceDiscoverer";
+  private Context mContext;
+  private WifiP2pManager mManager;
+  private Channel mChannel;
+  private IntentFilter mIntentFilter = new IntentFilter();
+  private boolean mIsRunning = false;
+  private BroadcastReceiver mReceiver = new BroadcastReceiver() {
+    @Override
+    public void onReceive(Context context, Intent intent) {
+      String action = intent.getAction();
+      if (WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION.equals(action)
+          && mManager != null) {
+        mManager.requestPeers(mChannel, mPeerListener);
+      }
+    }
+  };
+  private final PeerListListener mPeerListener = new PeerListListener() {
+    @Override
+    public void onPeersAvailable(WifiP2pDeviceList list) {
+      Log.d(TAG, list.toString());
+      for (WifiP2pDevice device : list.getDeviceList()) {
+        Utils.WifiDirectInfo info = Utils.parseWifiDirectName(device.deviceName);
+        if (info != null) {
+          String name = info.title;
+          int port = info.port;
+          reportUrlDevice(createUrlDeviceBuilder("WifiDirect" + name,
+              device.deviceAddress + ":" + port)
+            .setWifiAddress(device.deviceAddress)
+            .setWifiPort(port)
+            .setTitle(name)
+            .setDescription("")
+            .build());
+        }
+      }
+    }
+  };
+
+  public WifiUrlDeviceDiscoverer(Context context) {
+    mContext = context;
+  }
+
+  @Override
+  public synchronized void startScanImpl() {
+    mManager = (WifiP2pManager) mContext.getSystemService(Context.WIFI_P2P_SERVICE);
+    mChannel = mManager.initialize(mContext, mContext.getMainLooper(), null);
+    IntentFilter intentFilter = new IntentFilter();
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION);
+    mContext.registerReceiver(mReceiver, intentFilter);
+    mIsRunning = true;
+    discoverHelper(true);
+  }
+
+  public void discoverHelper(boolean retry) {
+    mManager.discoverPeers(mChannel, new WifiP2pManager.ActionListener() {
+      @Override
+      public void onSuccess() {
+      }
+
+      @Override
+      public void onFailure(int reasonCode) {
+        if (retry && reasonCode == 2) {
```

Comment:
Use the WifiP2pManager.BUSY constant here so we don't have to guess what 2 means